packages/lodestar-beacon-state-transition/src/fast/util/interface.ts

```diff
@@ -0,0 +1,15 @@
+/* eslint-disable @typescript-eslint/interface-name-prefix */
+import {IReadonlyEpochShuffling} from ".";
+import {ValidatorIndex, Slot} from "@chainsafe/lodestar-types";
+import {ByteVector} from "@chainsafe/ssz";
+
+/**
+ * Readonly interface for EpochContext.
+ */
+export interface ReadonlyEpochContext {
```

Comment:
you could add something like this:
```typescript
type Readonly<T> = {
    readonly [P in keyof T]: T[P];
}
export type ReadonlyEpochContext = ReadOnly<EpochContext>;
```

Reply:
yeah but I want to enforce `currentShuffling` to be `IReadonlyEpochShuffling` not `IEpochShuffling` ...

Reply:
You could create:
```typescript
type DeepReadonly<T> = {
    readonly [P in keyof T]: DeepReadonly<T[P]>;
}
```

Reply:
this way, everything should be readonly without extra interfaces and casting

Reply:
I failed to do that with this error
```
@chainsafe/lodestar:   Type 'DeepReadonly<(key: Vector<number>) => number>' has no call signatures.
@chainsafe/lodestar: 49     const index = this.chain.getEpochContext().pubkey2index.get(pubkey);
```

also, I want to use ReadonlyMap<> instead of Readonly<Map<>> to avoid write operations to the map.